core/api/src/main/scala/org/virtuslab/ideprobe/protocol/ProjectRef.scala

```diff
@@ -1,10 +1,42 @@
 package org.virtuslab.ideprobe.protocol
 
+import com.typesafe.config.ConfigValueFactory
+import org.virtuslab.ideprobe.ConfigFormat
+import pureconfig.error.{CannotConvert, ConfigReaderFailures, ConvertFailure}
+import pureconfig.generic.semiauto.deriveReader
+import pureconfig.{ConfigCursor, ConfigReader, ConfigWriter}
+
 sealed trait ProjectRef
 
-object ProjectRef {
+object ProjectRef extends ConfigFormat {
   case object Default extends ProjectRef
   case class ByName(name: String) extends ProjectRef
 
   def apply(name: String): ProjectRef = ByName(name)
+
+  private val defaultProjectRefConfigReader: ConfigReader[ProjectRef.Default.type] = (cursor: ConfigCursor) => {
+    cursor.asString match {
+      case Left(failures) => Left[ConfigReaderFailures, ProjectRef.Default.type](failures)
```

Comment:
Step 1. You don't need to provide types for Left and Right, by the type of val in line 17 it is known what kind of Either is expected as return type. Compiler will just check if it seems ok. 
Step 2. left is noop, you don't change anything, we just want to transform Right, so you can just use `map` instead of pattern matching. Actually `flatMap`, as you might return Left or Right.
Step 3. use built in factory function:
```
ConfigReader.fromStringOpt(s => if (s.toLowerCase == "default") Some(Default) else None)
```